src/Microsoft.Extensions.FileProviders.Physical/PollingWildCardChangeToken.cs

```diff
@@ -0,0 +1,153 @@
+            Clock = clock;
+
+            _matcher = new Matcher(StringComparison.OrdinalIgnoreCase);
+            _matcher.AddInclude(pattern);
+            CalculateChanges();
+        }
+
+        /// <inheritdoc />
+        public bool ActiveChangeCallbacks => false;
+
+        public IClock Clock { get; }
+
+        /// <inheritdoc />
+        public bool HasChanged
+        {
+            get
+            {
+                if (_changed)
+                {
+                    return _changed;
+                }
+
+                if (Clock.UtcNow - _lastScanTimeUtc >= PollingInterval)
+                {
+                    lock (_enumerationLock)
+                    {
+                        _changed = CalculateChanges();
+                    }
+                }
+
+                _lastScanTimeUtc = Clock.UtcNow;
+                return _changed;
+            }
+        }
+
+        private bool CalculateChanges()
+        {
+            var result = _matcher.Execute(_directoryInfo);
+            var hasChanges = false;
+
+            var files = result.Files.OrderBy(f => f.Path, StringComparer.OrdinalIgnoreCase);
+#if NET451
+            using (var sha256 = new IncrementalHash())
+#else
+            using (var sha256 = IncrementalHash.CreateHash(HashAlgorithmName.SHA256))
+#endif
+            {
+                foreach (var file in files)
+                {
+                    if (!hasChanges &&
+                        _lastScanTimeUtc < GetLastWriteUtc(file.Path))
+                    {
+                        // _lastScanTimeUtc is the greatest timestamp that any last writes could have been.
+                        // If a file has a newer timestamp than this value, it must've changed.
+                        hasChanges = true;
+                    }
+
+                    ComputeHash(sha256, file.Path);
+                }
+
+                var currentHash = sha256.GetHashAndReset();
+                hasChanges |= !ArrayEquals(_previousHash, currentHash);
+                _previousHash = currentHash;
+            }
+
+            return hasChanges;
+        }
+
+        protected virtual DateTime GetLastWriteUtc(string path)
+        {
+            return File.GetLastWriteTimeUtc(Path.Combine(_directoryInfo.FullName, path));
+        }
+
+        private bool ArrayEquals(byte[] _previousHash, byte[] currentHash)
+        {
+            if (_previousHash == null)
+            {
+                // First run
+                return true;
+            }
+
+            Debug.Assert(_previousHash.Length == currentHash.Length);
+            for (var i = 0; i < _previousHash.Length; i++)
+            {
+                if (_previousHash[i] != currentHash[i])
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private void ComputeHash(IncrementalHash sha256, string path)
+        {
+            var byteCount = Encoding.UTF8.GetByteCount(path);
```

Comment:
My original thinking was .NET stores strings as UTF-16 (Unicode) so you would think its cheaper. But, it looks like the [implementation](http://referencesource.microsoft.com/#mscorlib/system/text/unicodeencoding.cs,2b03a0eac68f75f2) of `UnicodeEncoding.GetBytes` isn't much better that `UTF8Encoding.GetBytes`. It's be fastest to do this in unsafe code and use the string's raw memory (assuming the sha hash doesn't manipulate the original memory).